packages/transaction-controller/src/helpers/MultichainHelper.ts

```diff
@@ -0,0 +1,401 @@
+ * @property includeTokenTransfers - Whether or not to include ERC20 token transfers.
+ * @property isEnabled - Whether or not incoming transaction retrieval is enabled.
+ * @property queryEntireHistory - Whether to initially query the entire transaction history or only recent blocks.
+ * @property updateTransactions - Whether to update local transactions using remote transaction data.
+ */
+type IncomingTransactionOptions = {
+  includeTokenTransfers?: boolean;
+  isEnabled?: () => boolean;
+  queryEntireHistory?: boolean;
+  updateTransactions?: boolean;
+};
+
+type Trackers = {
+  nonceTracker: NonceTracker;
+  pendingTransactionTracker: PendingTransactionTracker;
+  incomingTransactionHelper: IncomingTransactionHelper;
+};
+
+export class MultichainHelper {
+  #isMultichainEnabled: boolean;
+
+  readonly #ethQuery: EthQuery;
+
+  readonly #nonceTracker: NonceTracker;
+
+  readonly #incomingTransactionOptions: IncomingTransactionOptions;
+
+  readonly #findNetworkClientIdByChainId: NetworkController['findNetworkClientIdByChainId'];
+
+  readonly #getNetworkClientById: NetworkController['getNetworkClientById'];
+
+  readonly #getNetworkClientRegistry: NetworkController['getNetworkClientRegistry'];
+
+  readonly #removeIncomingTransactionHelperListeners: (
+    IncomingTransactionHelper: IncomingTransactionHelper,
+  ) => void;
+
+  readonly #removePendingTransactionTrackerListeners: (
+    pendingTransactionTracker: PendingTransactionTracker,
+  ) => void;
+
+  readonly #createTrackersForNetworkClient: (
+    networkClient: NetworkClient,
+    etherscanRemoteTransactionSource: EtherscanRemoteTransactionSource,
+  ) => Trackers;
+
+  readonly #acquireNonceLockForChainIdKey: (opts: {
+    chainId: Hex;
+    key?: string;
+  }) => Promise<() => void>;
+
+  readonly #trackingMap: Map<NetworkClientId, Trackers> = new Map();
+
+  readonly etherscanRemoteTransactionSourcesMap: Map<
+    Hex,
+    EtherscanRemoteTransactionSource
+  > = new Map();
+
+  constructor({
+    isMultichainEnabled,
+    ethQuery,
+    nonceTracker,
+    incomingTransactionOptions,
+    findNetworkClientIdByChainId,
+    getNetworkClientById,
+    getNetworkClientRegistry,
+    removeIncomingTransactionHelperListeners,
+    removePendingTransactionTrackerListeners,
+    createTrackersForNetworkClient,
+    acquireNonceLockForChainIdKey,
+  }: {
+    isMultichainEnabled: boolean;
+    ethQuery: EthQuery;
+    nonceTracker: NonceTracker;
+    incomingTransactionOptions: IncomingTransactionOptions;
+
+    findNetworkClientIdByChainId: NetworkController['findNetworkClientIdByChainId'];
+    getNetworkClientById: NetworkController['getNetworkClientById'];
+    getNetworkClientRegistry: NetworkController['getNetworkClientRegistry'];
+
+    removeIncomingTransactionHelperListeners: (
+      IncomingTransactionHelper: IncomingTransactionHelper,
+    ) => void;
+    removePendingTransactionTrackerListeners: (
+      pendingTransactionTracker: PendingTransactionTracker,
+    ) => void;
```

Comment:
is this the responsibility of the TransactionController or this helper? Where should these methods live

Reply:
Since the global instances don't need to be removed, this feels a specific problem to the `MultichainHelper` so we can encapsulate this within it.

Reply:
I think this should stay as is. I agree that the global instances don't need to be removed, but for the purposes of testing, the listeners on the global instances should be cleared when destroy() is called (which likely will only ever be called in tests).

The TransactionController essentially helps orchestrate high level transaction flows by glueing the helpers and utils together. Adding listeners is an example of connecting a helper to a higher level flow, so add listener methods should stay in TransactionController. Given than, it makes sense for removal of listeners to also live there.